libs/core/src/test/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoaderTests.java

```diff
@@ -0,0 +1,459 @@
+            EmbeddedImplClassLoaderTests.class.getClassLoader()
+        );
+
+        EmbeddedImplClassLoader loader = EmbeddedImplClassLoader.getInstance(parent, "x-foo");
+        Class<?> c = loader.loadClass("p.FooBar");
+        return c.getConstructor().newInstance();
+    }
+
+    public void testCompoundEnumeration() {
+        @SuppressWarnings("unchecked")
+        var enumerations = (Enumeration<String>[]) new Enumeration<?>[] {
+            Collections.enumeration(List.of("hello")),
+            Collections.enumeration(List.of("world")) };
+        var compoundEnumeration = new CompoundEnumeration<>(enumerations);
+
+        List<String> l = new ArrayList<>();
+        while (compoundEnumeration.hasMoreElements()) {
+            l.add(compoundEnumeration.nextElement());
+        }
+        assertThat(l, contains(is("hello"), is("world")));
+        expectThrows(NoSuchElementException.class, () -> compoundEnumeration.nextElement());
+    }
+
+    /* Basic test for resource loading. */
+    public void testResourcesBasic() throws Exception {
+        Path topLevelDir = createTempDir();
+
+        Map<String, byte[]> jarEntries = new HashMap<>();
+        jarEntries.put("IMPL-JARS/res/LISTING.TXT", "res-impl.jar".getBytes(UTF_8));
+        jarEntries.put("IMPL-JARS/res/res-impl.jar/p/res.txt", "Hello World".getBytes(UTF_8));
+
+        Path outerJar = topLevelDir.resolve("impl.jar");
+        JarUtils.createJarWithEntries(outerJar, jarEntries);
+        URLClassLoader parent = URLClassLoader.newInstance(
+            new URL[] { outerJar.toUri().toURL() },
+            EmbeddedImplClassLoaderTests.class.getClassLoader()
+        );
+
+        EmbeddedImplClassLoader loader = EmbeddedImplClassLoader.getInstance(parent, "res");
+        URL url = loader.findResource("p/res.txt");
+        assertThat(url.toString(), endsWith("impl.jar!/IMPL-JARS/res/res-impl.jar/p/res.txt"));
+
+        var resources = loader.findResources("p/res.txt");
+        assertThat(Collections.list(resources), contains(hasToString(endsWith("impl.jar!/IMPL-JARS/res/res-impl.jar/p/res.txt"))));
+    }
+
+    /*
+     * Tests that the root version of a resource is located, when the multi-release attribute is absent.
+     */
+    public void testResourcesWithoutMultiRelease() throws Exception {
+        assumeTrue("JDK version not greater than or equal to 17", Runtime.version().feature() >= 17);
+        testResourcesVersioned(false, 0, 9);
+        testResourcesVersioned(false, 0, 15);
+        testResourcesVersioned(false, 0, 17);
+        testResourcesVersioned(false, 0, 11, 10, 9, 8);
+    }
+
+    /*
+     * Tests that the specific, 9, version of a resource is located, when the multi-release attribute
+     * is present and the versioned entry is less than or equal to the runtime version.
+     */
+    public void testResourcesVersioned9() throws Exception {
+        assumeTrue("JDK version not greater than or equal to 9", Runtime.version().feature() >= 9);
```

Comment:
Do we need these assumes since the build is on jdk 17+?

Reply:
We probably done need them, but they serve mostly as a comment to the reader.